dsindex/indexer.go

```diff
@@ -0,0 +1,207 @@
+	ForEach(index string, fn func(index, id string) bool) error
+
+	// HasKey determines the specified index contains the specified primary key
+	HasKey(index, id string) (bool, error)
+
+	// HasAny determines if any key is in the specified index.  If index is
+	// empty string, then all indexes are searched.
+	HasAny(index string) (bool, error)
+
+	// Search returns all keys for the given index
+	Search(index string) (ids []string, err error)
+
+	// Synchronize the indexes in this Indexer to match those of the given
+	// Indexer. The indexPath prefix is not synchronized, only the index/key
+	// portion of the indexes.
+	SyncTo(reference Indexer) (changed bool, err error)
+}
+
+// indexer is a simple implementation of Indexer.  This implementation relies
+// on the underlying data store supporting efficent querying by prefix.
+//
+// TODO: Consider adding caching
+type indexer struct {
+	dstore    ds.Datastore
+	indexPath string
+}
+
+// New creates a new datastore index.  All indexes are stored prefixed with the
+// specified index path.
+func New(dstore ds.Datastore, indexPath string) Indexer {
+	return &indexer{
+		dstore:    dstore,
+		indexPath: indexPath,
+	}
+}
+
+func (x *indexer) Add(index, id string) error {
+	key := ds.NewKey(path.Join(x.indexPath, index, id))
+	return x.dstore.Put(key, []byte{})
+}
+
+func (x *indexer) Delete(index, id string) error {
+	return x.dstore.Delete(ds.NewKey(path.Join(x.indexPath, index, id)))
+}
+
+func (x *indexer) DeleteAll(index string) (int, error) {
+	ents, err := x.queryPrefix(path.Join(x.indexPath, index))
+	if err != nil {
+		return 0, err
+	}
+
+	for i := range ents {
+		err = x.dstore.Delete(ds.NewKey(ents[i].Key))
+		if err != nil {
+			return 0, err
+		}
+	}
+
+	return len(ents), nil
+}
+
+func (x *indexer) ForEach(index string, fn func(idx, id string) bool) error {
+	q := query.Query{
+		Prefix:   path.Join(x.indexPath, index),
+		KeysOnly: true,
+	}
+	results, err := x.dstore.Query(q)
+	if err != nil {
+		return err
+	}
+
+	for {
+		r, ok := results.NextSync()
+		if !ok {
+			break
+		}
+		if r.Error != nil {
+			err = r.Error
+			break
+		}
+
+		ent := r.Entry
+		if !fn(path.Base(path.Dir(ent.Key)), path.Base(ent.Key)) {
+			break
+		}
+	}
+	results.Close()
+
+	return err
+}
+
+func (x *indexer) HasKey(index, id string) (bool, error) {
+	return x.dstore.Has(ds.NewKey(path.Join(x.indexPath, index, id)))
+}
+
+func (x *indexer) HasAny(index string) (bool, error) {
+	var any bool
+	err := x.ForEach(index, func(idx, id string) bool {
+		any = true
+		return false
+	})
+	return any, err
+}
+
+func (x *indexer) Search(index string) ([]string, error) {
+	ents, err := x.queryPrefix(path.Join(x.indexPath, index))
+	if err != nil {
+		return nil, err
+	}
+	if len(ents) == 0 {
+		return nil, nil
+	}
+
+	ids := make([]string, len(ents))
+	for i := range ents {
+		ids[i] = path.Base(ents[i].Key)
+	}
+	return ids, nil
+}
+
+func (x *indexer) SyncTo(ref Indexer) (bool, error) {
+	// Build reference index map
```

Comment:
Is this really part of the interface or is this just a utility function that can work on any implementation of Indexer?

Reply:
The latter.  Made it part of the interface to force implementations implemented it.

Reply:
But do implementations need to implement it? This function doesn't rely on any internal parameters of `indexer` , the references to `x.dstore.Put/Delete` can be replaced by Indexer's Add and Delete functions.

That would make this function into something like `SyncIndex(ref, target Indexer) (bool, error)`

Reply:
Done. Removed from interface.
Ah, thank you for seeing this.